File: frontend/src/interfaces/Bookingstore.ts
import { PetInterface } from "./Pet";


export interface BookingInterface {
    pets: PetInterface[];       // รองรับเป็นอาร์เรย์ของ PetInterface
    BookerUser: any;            // ข้อมูลผู้จอง
    ID?: number;                // ID ของการจอง (ถ้ามี)
    booker_user_id: number;     // ID ของผู้จอง
    store_id: number;           // ID ของร้าน
    service_id: number;
    date: string;               // วันที่จอง (ในรูปแบบ ISO)
    booking_time: string;       // เวลาในการจอง
    status?: string;            // สถานะ (ถ้ามี)
    notes?: string;             // หมายเหตุ (ถ้ามี)
    total_cost: number;         // ค่าใช้จ่ายทั้งหมด
    contact_number: string;     // หมายเลขโทรศัพท์ของผู้จอง
    count_pet: number;          // จำนวนสัตว์เลี้ยงที่จอง

    User?: any;                 // ข้อมูลผู้ใช้ (ถ้ามี)
    Store?: any;                // ข้อมูลร้าน (ถ้ามี)
    Service?: {                 // ข้อมูลบริการ (ถ้ามี)
        name_service: string;   // ชื่อบริการ
        price: number;          // ราคา
        duration: number;       // ระยะเวลา
    };
}

File: backend/entity/bookingstore.go
package entity

import (
	"gorm.io/gorm"
	"time"
)

type Bookingstore struct {
	gorm.Model

	BookerUserID uint  `json:"booker_user_id" gorm:"index" valid:"required~BookerUserID is required"`
	BookerUser   Users `gorm:"foreignKey:BookerUserID;references:ID"`

	StoreID uint  `json:"store_id" gorm:"index" valid:"required~StoreID is required"`
	Store   Store `gorm:"foreignKey:StoreID;references:ID"`

	ServiceID uint    `json:"service_id" valid:"required~ServiceID is required"`
	Service   Service `gorm:"foreignKey:ServiceID;references:ID"`

	BookingTime string `json:"booking_time" valid:"required~BookingTime is required"`

	Date time.Time `json:"date" valid:"required~Date is required"`

	Status string `json:"status" valid:"in(confirmed|pending|cancelled)~Status must be 'confirmed', 'pending', or 'cancelled'"`

	Notes string `json:"notes" gorm:"type:text"`

	TotalCost float32 `json:"total_cost" valid:"required~TotalCost is required"`

	ContactNum string `json:"contact_num" valid:"required~Contact number is required,matches(^\\d{10}$)~Contact number must be 10 digits"`

	CountPet int `json:"count_pet" valid:"required~CountPet is required"`

	Pets []Pet `gorm:"many2many:booking_pets;joinForeignKey:BookingID;joinReferences:PetID;" json:"pets"`
}
